docs/user-guide/configuring-workers.rst
##########################################
Configuring Noteburst's JupyterLab workers
##########################################

Noteburst works by operating a cluster of workers that each manages their own JupyterLab pods.
This page describes how these workers are configured.

Background: Kubernetes architecture
===================================

In Kubernetes, the workers are deployed as a Kubernetes Deployment.
A Deployment enables multiple Noteburst worker instances to run at the same time.
These workers share the same configuration (typically through a Kubernetes ConfigMap).
This means that individual workers can't be assigned specific RSP/JupyterLab user accounts.
Noteburst works around this by configuring the deployment of Noteburst workers with a pool of identities.
When a worker starts up, it picks an available identity from the pool and uses that identity to run the JupyterLab pod.
See the next section for details.

.. _worker-identities-yaml:

Worker identities
=================

Each Noteburst worker pod runs a JupyterLab server under a specific, and unique, identity.
These identities are bot accounts.
In some RSP environments, such as the USDF, these identities need to be associated with actual user accounts.

When a Noteburst worker pod starts up, it picks an available identity from a pool of available identities.
These identifies are configured in a file, that path of which is specified with :envvar:`NOTEBURST_WORKER_IDENTITIES_PATH`.
This file is a YAML-formatted file that looks like this:

.. code-block:: yaml
   :caption: identities.yaml

   - username: "bot-noteburst00"
   - username: "bot-noteburst01"
   - username: "bot-noteburst02"
   - username: "bot-noteburst03"
   - username: "bot-noteburst04"
   - username: "bot-noteburst05"

The YAML file consists of a list of identities.
At a minimum, an identity requires a ``username`` field.

In some environments where Gafaelfawr cannot provide a uid for a user, a ``uid`` must be specified:

.. code-block:: yaml
   :caption: identities.yaml

   - username: "bot-noteburst00"
     uid: 90000
   - username: "bot-noteburst01"
     uid: 90001
   - username: "bot-noteburst02"
     uid: 90002
   - username: "bot-noteburst03"
     uid: 90003
   - username: "bot-noteburst04"
     uid: 90004
   - username: "bot-noteburst05"
     uid: 90005


docs/user-guide/environment-variables.rst
#####################
Environment variables
#####################

Noteburst uses environment variables for configuration.
In practice, these variables are typically set as Helm values and 1Password/Vault secrets that are injected into the container as environment variables.
See the `Phalanx documentation for Noteburst <https://phalanx.lsst.io/applications/noteburst/index.html>`__ for more information on the Phalanx-specific configurations.

.. envvar:: SAFIR_NAME

   (string, default: "Noteburst") The name of the application.
   This is used in the metadata endpoint.

.. envvar:: SAFIR_PROFILE

   (string enum: "production" [default], "development") The application run profile.
   Use production to enable JSON structured logging.

.. envvar:: SAFIR_LOG_LEVEL

   (string enum: "debug", "info" [default], "warning", "error", "critical") The application log level.

.. envvar:: NOTEBURST_PATH_PREFIX
    
   (string, default: "/noteburst") The path prefix for the Noteburst application.
   This is used to configure the application's URL.

.. envvar:: NOTEBURST_ENVIRONMENT_URL

   (string) The base URL of the Rubin Science Platform environment.
   This is used for creating URLs to services, such as JupyterHub.

.. envvar:: NOTEBURST_JUPYTERHUB_PATH_PREFIX

   (string, default: "/nb") The path prefix for the JupyterHub application.

.. envvar:: NOTEBURST_NUBLADO_CONTROLLER_PATH_PREFIX

   (string, default: "/nublado") The path prefix for the Nublado controller service.

.. envvar:: NOTEBURST_GAFAELFAWR_TOKEN

   (secret string) This token is used to make an admin API call to Gafaelfawr to get a token for the user.

.. envvar:: NOTEBURST_REDIS_URL

   (string) The URL of the Redis server, used by the worker queue.

.. envvar:: NOTEBURST_ARQ_MODE

   (string enum: "production" [default], "test") The Arq worker mode.
   The production mode uses the Redis server, while the test mode mocks queue interactions for testing the application.

.. envvar:: NOTEBURST_WORKER_IDENTITIES_PATH

   (string) The path to the Science Platform worker identities file.
   See :ref:`worker-identities-yaml`.

.. envvar:: NOTEBURST_WORKER_QUEUE_NAME

   (string) The name of arq queue the workers process.

.. envvar:: NOTEBURST_WORKER_LOCK_REDIS_URL

   (Redis URL) The URL of the Redis server, used by the worker lock.

.. envvar:: NOTEBURST_WORKER_JOB_TIMEOUT

   (integer, default: 3000) The timeout for a worker job, in seconds.

.. envvar:: NOTEBURST_WORKER_TOKEN_LIFETIME

   (integrer, default: 2419200) The worker auth token lifetime in seconds.

.. envvar:: NOTEBURST_WORKER_TOKEN_SCOPES

   (string, default: "exec:notebook") The worker (nublado pod) token scopes, as a comma-separated string.

.. envvar:: NOTEBURST_WORKER_IMAGE_SELECTOR

   (string enum: "recommended" [default], "weekly", "reference") The method for selecting a Jupyter image to run.
   For "reference" see :envvar:`NOTEBURST_WORKER_IMAGE_REFERENCE`.

.. envvar:: NOTEBURST_WORKER_IMAGE_REFERENCE

   (string) The tag of the Jupyter image to run. This is used when :envvar:`NOTEBURST_WORKER_IMAGE_SELECTOR` is set to "reference".

.. envvar:: NOTEBURST_WORKER_KEEPALIVE

   (string, enum: "normal" [default], "fast", "disabled") The worker keep alive mode.
   The regular keep-alive execises the JupyterLab pod every 5 minutes. The fast mode exercises the pod every 30 seconds.
   The disabled mode does not exercise the pod.


CHANGELOG.md
# Change log

## 0.7.1 (2023-07-23)

Fixes:

- Add additional logging of JupyterLab spawning failures in workers.

Other changes:

- Added documentation for configuration environment variables.
- Added OpenAPI docs, rendered by Redoc, to the Sphinx documentation site.

## 0.7.0 (2023-05-22)

- The JupyterHub service's URL path prefix is now configurable with the `NOTEBURST_JUPYTERHUB_PATH_PREFIX` environment variable. The default is `/nb/`, which is the existing value.
- The Nublado JupyterLab Controller service's URL path prefix is configurable with the `NOTEBURST_NUBLADO_CONTROLLER_PATH_PREFIX` environment variable. The default is `/nublado`, which is the existing value.

## 0.6.3 (2023-04-20)

- Fix how failed notebook executions are handled. Previously failed notebooks would prevent Noteburst from getting the results of the execution job. Now the job is shown as concluded but unsuccessful by the `/v1/notebooks/{job_id}` endpoint.
- Structure uvicorn server logging.

## 0.6.2 (2023-04-12)

- Stop following redirects from the `hub/login` endpoint.
- Explicitly shut down the lab pod on worker shutdown.
- Update dependencies.

## 0.6.1 (2023-03-28)

- Additional updates for JupyterLab Controller image API endpoint.

## 0.6.0 (2023-02-16)

- Migrated from the Cachemachine API to the new JupyterLab Controller API for obtaining the list of available Docker images for JupyterLab workers.
- Migrated to Python 3.11
- Adopted pyproject.toml for project metadata and dropped setup.cfg.

## 0.5.0 (2022-07-04)

- Its now possible to skip retries on notebook execution failures in the `nbexec` task by passing an `enable_retry=False` keyword argument.
  This is useful for applications that use Noteburst for continuous integration.
- Updated dependencies.

## 0.4.0 (2022-06-15)

- The worker identity configuration can now omit the `uid` field for environments where Gafaelfawr is able to assign a UID (e.g. through an LDAP backend).
- New configurations for workers:
  - The new `NOTEBURST_WORKER_TOKEN_LIFETIME` environment variable enables you to configure the lifetime of the workers' authentication tokens. The default matches the existing behavior, 28 days.
  - `NOTEBURST_WORKER_TOKEN_SCOPES` environment variable enables you to set what token scopes the nublado2 bot users should have, as a comma-separated list.
  - `NOTEBURST_WORKER_IMAGE_SELECTOR` allows you to specify what stream of Nublado image to select. Can be `recommended`, `weekly` or `reference`. If the latter, you can specify the specific Docker image with `NOTEBURST_WORKER_IMAGE_REFERENCE`.
  - The `NOTEBURST_WORKER_KEEPALIVE` configuration controls whether the worker keep alive function is run (to defeat the Nublado pod culler), and at what frequency. Set to `disabled` to disable; `fast` to run every 30 seconds; or `normal` to run every 5 minutes.
- Noteburst now uses the arq client and dependency from Safir 3.2, which was originally developed from Noteburst.

## 0.3.0 (2022-05-24)

Improved handling of the JupyterLab pod for noteburst workers:

- If the JupyterLab pod goes away (such as if it is culled), the Noteburst workers shuts down so that Kubernetes creates a new worker with a new JupyterLab pod. A lost JupyterLab pod is detected by a 400-class response when submitting a notebook for execution.

- If a worker starts up and a JupyterLab pod already exists for an unclaimed identity, the noteburst worker will continue to cycle through available worker identities until the JupyterLab start up is successful. This handles cases where a Noteburst worker restarts, but the JupyterLab pod did not shut down and thus is "orphaned."

- Each JupyterLab worker runs a "keep alive" function that exercises the JupyterLab pod's Python kernel. This is meant to counter the "culler" that deletes dormant JupyterLab pods in the Rubin Science Platform. Currently the keep alive function runs every 30 seconds.

- The default arq job execution timeout is now configurable with the `NOTEBURST_WORKER_JOB_TIMEOUT` environment variable. By default it is 300 seconds (5 minutes).

## 0.2.0 (2022-03-14)

- Initial version of the `/v1/` HTTP API.
- Migration to Safir 3 and its database framework.
- Noteburst is now cross-published to the GitHub Container Registry, `ghcr.io/lsst-sqre/noteburst`.
- Migration to Python 3.10.

## 0.1.0 (2021-09-29)

- Initial development version of Noteburst.


src/noteburst/config.py
"""Configuration definition."""

from __future__ import annotations

from enum import Enum
from pathlib import Path
from typing import Any, Mapping, Optional
from urllib.parse import urlparse

from arq.connections import RedisSettings
from pydantic import (
    BaseSettings,
    Field,
    HttpUrl,
    RedisDsn,
    SecretStr,
    validator,
)
from safir.arq import ArqMode

__all__ = ["Config", "Profile", "LogLevel"]


class Profile(str, Enum):
    production = "production"

    development = "development"


class LogLevel(str, Enum):
    DEBUG = "DEBUG"

    INFO = "INFO"

    WARNING = "WARNING"

    ERROR = "ERROR"

    CRITICAL = "CRITICAL"


class JupyterImageSelector(str, Enum):
    """Possible ways of selecting a JupyterLab image."""

    recommended = "recommended"
    """Currently recommended image."""

    weekly = "weekly"
    """Current weekly image."""

    reference = "reference"
    """Select a specific image by reference."""


class WorkerKeepAliveSetting(str, Enum):
    """Modes for the worker keep-alive function."""

    disabled = "disabled"
    """Do not run a keep-alive function."""

    fast = "fast"
    """Run the keep-alive function at a high frequency (every 30 seconds)."""

    normal = "normal"
    """Run the keep-alive function at a slower frequency (i.e. 5 minutes)."""


class Config(BaseSettings):
    name: str = Field("Noteburst", env="SAFIR_NAME")

    profile: Profile = Field(Profile.production, env="SAFIR_PROFILE")

    log_level: LogLevel = Field(LogLevel.INFO, env="SAFIR_LOG_LEVEL")

    logger_name: str = "noteburst"
    """The root name of the Python logger, which is also the name of the
    root Python module.
    """

    path_prefix: str = Field("/noteburst", env="NOTEBURST_PATH_PREFIX")
    """The URL path prefix where noteburst is hosted."""

    environment_url: HttpUrl = Field(env="NOTEBURST_ENVIRONMENT_URL")
    """The base URL of the Rubin Science Platform environment.

    This is used for creating URLs to services, such as JupyterHub.
    """

    jupyterhub_path_prefix: str = Field(
        "/nb/",
        env="NOTEBURST_JUPYTERHUB_PATH_PREFIX",
        description="The path prefix for the JupyterHub service.",
    )

    nublado_controller_path_prefix: str = Field(
        "/nublado",
        env="NOTEBURST_NUBLADO_CONTROLLER_PATH_PREFIX",
        description="The path prefix for the Nublado controller service.",
    )

    gafaelfawr_token: SecretStr = Field(env="NOTEBURST_GAFAELFAWR_TOKEN")
    """This token is used to make an admin API call to Gafaelfawr to get a
    token for the user.
    """

    redis_url: RedisDsn = Field(
        env="NOTEBURST_REDIS_URL",
        # Preferred by mypy over a string default
        default_factory=lambda: RedisDsn("redis://localhost:6379/1"),
    )
    """URL for the redis instance, used by the worker queue."""

    arq_mode: ArqMode = Field(ArqMode.production, env="NOTEBURST_ARQ_MODE")

    @property
    def arq_redis_settings(self) -> RedisSettings:
        """Create a Redis settings instance for arq."""
        url_parts = urlparse(self.redis_url)
        redis_settings = RedisSettings(
            host=url_parts.hostname or "localhost",
            port=url_parts.port or 6379,
            database=int(url_parts.path.lstrip("/")) if url_parts.path else 0,
        )
        return redis_settings


class WorkerConfig(Config):
    identities_path: Path = Field(..., env="NOTEBURST_WORKER_IDENTITIES_PATH")
    """Path to the configuration file with the pool of Science Platform
    identities available to workers.
    """

    queue_name: str = Field("arq:queue", env="NOTEBURST_WORKER_QUEUE_NAME")
    """Name of the arq queue that the worker processes from."""

    identity_lock_redis_url: RedisDsn = Field(
        env="NOTEBURST_WORKER_LOCK_REDIS_URL",
        # Preferred by mypy over a string default
        default_factory=lambda: RedisDsn("redis://localhost:6379/1"),
    )

    job_timeout: int = Field(
        300,
        env="NOTEBURST_WORKER_JOB_TIMEOUT",
        description=(
            "The timeout, in seconds, for a job until it is timed out."
        ),
    )

    worker_token_lifetime: int = Field(
        2419200,
        env="NOTEBURST_WORKER_TOKEN_LIFETIME",
        description="Worker auth token lifetime in seconds.",
    )

    worker_token_scopes: str = Field(
        "exec:notebook",
        env="NOTEBURST_WORKER_TOKEN_SCOPES",
        description=(
            "Worker (nublado2 pod) token scopes as a comma-separated string."
        ),
    )

    image_selector: JupyterImageSelector = Field(
        JupyterImageSelector.recommended,
        env="NOTEBURST_WORKER_IMAGE_SELECTOR",
        description="Method for selecting a Jupyter image to run.",
    )

    image_reference: Optional[str] = Field(
        None,
        env="NOTEBURST_WORKER_IMAGE_REFERENCE",
        description=(
            "Docker image reference, if NOTEBURST_WORKER_IMAGE_SELECTOR is "
            "``reference``."
        ),
    )

    worker_keepalive: WorkerKeepAliveSetting = Field(
        WorkerKeepAliveSetting.normal, env="NOTEBURST_WORKER_KEEPALIVE"
    )

    @property
    def aioredlock_redis_config(self) -> list[str]:
        """Redis configurations for aioredlock."""
        return [str(self.identity_lock_redis_url)]

    @validator("image_reference")
    def is_image_ref_set(
        cls, v: Optional[str], values: Mapping[str, Any]
    ) -> Optional[str]:
        """Validate that image_reference is set if image_selector is
        set to reference.
        """
        if (
            v is None
            and values["image_selector"] == JupyterImageSelector.reference
        ):
            raise ValueError(
                "Set NOTEBURST_WORKER_IMAGE_REFERENCE since "
                "NOTEBURST_WORKER_IMAGE_SELECTOR is ``reference``."
            )

        return v

    @property
    def parsed_worker_token_scopes(self) -> list[str]:
        """Sequence of worker token scopes, parsed from the comma-separated
        list in `worker_token_scopes`.
        """
        return [t.strip() for t in self.worker_token_scopes.split(",") if t]


config = Config()
"""Configuration for noteburst."""
